### ISSUE TITLE
Unexpected import reserved word

### ISSUE THREAD

**Issue by rickmed** [gadicc/meteor-hmr#94]
Using gadicc:ecmascript-hot@2.0.0-beta.3

Error:
```
W20160521-19:04:32.500(-4)? (STDERR) import { Mongo } from 'meteor/mongo';
W20160521-19:04:32.505(-4)? (STDERR) ^^^^^^
W20160521-19:04:32.511(-4)? (STDERR) SyntaxError: Unexpected reserved word
```

on running meteor after installing gadicc:ecmascript-hot@2.0.0-beta.3:
```
Changes to your project's package version selections:

gadicc:babel-compiler-babelrc  upgraded from 6.6.4_2 to 6.6.4_5
gadicc:ecmascript-hot*         upgraded from 1.3.2-refactor.12 to 2.0.0-beta.3
gadicc:hot*+                   upgraded from 0.0.25 to 2.0.0-beta.1
gadicc:hot-build*+             upgraded from 0.0.14 to 2.0.0-beta.3
gadicc:modules-runtime-hot     upgraded from 0.6.3_6 to 0.6.3_8
gadicc:package-json            added, version 1.0.4
package-version-parser         added, version 3.0.9-beta.0
sanjo:meteor-files-helpers     added, version 1.2.0_1
sanjo:meteor-version           added, version 1.0.0


* These packages have been updated to new versions that are not backwards
compatible.

+ In order to resolve constraints, we had to use experimental versions of these
packages.
```

**Comment by tomitrescak** [comment]
I had a same problem .. I copy pasted another project and got over the error. Was trying to understand what caused the issue, but no luck. But yea, the error is there ;(

**Comment by sammkj** [comment]
Can you guys try adding `babel-preset-es2015`? The error disappeared after adding that.

**Comment by rickmed** [comment]
@sammkj what would be your babelrc?

**Comment by sammkj** [comment]
my current babelrc looks like this:

```
{
  "presets": [
    "meteor",
    "es2015",
    "react"
  ],
  "plugins": [....]
}
```

**Comment by rickmed** [comment]
@sammkj thanks I got passed that issue but know I'm getting in the browser console this when updating a file:
```
warning.js:44 Warning: React.createElement: type should not be null, undefined, boolean, or number. It should be a string (for DOM elements) or a ReactClass (for composite components).warning @ warning.js:44ReactElementValidator.createElement @ ReactElementValidator.js:221patchedCreateElement @ patch.dev.js:27renderApp @ Index.jsx:21(anonymous function) @ Index.jsx:38canAccept @ hot.js:207requirersUntil @ hot.js:105requirersUntil @ hot.js:110requirersUntil @ hot.js:110hotWalker @ hot.js:162walkFileTree @ modules-runtime.js?hash=5e943de…&hot=1:477walkFileTree @ modules-runtime.js?hash=5e943de…&hot=1:481walkFileTree @ modules-runtime.js?hash=5e943de…&hot=1:481walkFileTree @ modules-runtime.js?hash=5e943de…&hot=1:481meteorInstallHot @ hot.js:143(anonymous function) @ hot.js?hash=d34b0a8…:1
hot.js:120 [gadicc:hot] An error occured trying to accept hmr for /imports/client/Index.jsxlogAndFail @ hot.js:120canAccept @ hot.js:211requirersUntil @ hot.js:105requirersUntil @ hot.js:110requirersUntil @ hot.js:110hotWalker @ hot.js:162walkFileTree @ modules-runtime.js?hash=5e943de…&hot=1:477walkFileTree @ modules-runtime.js?hash=5e943de…&hot=1:481walkFileTree @ modules-runtime.js?hash=5e943de…&hot=1:481walkFileTree @ modules-runtime.js?hash=5e943de…&hot=1:481meteorInstallHot @ hot.js:143(anonymous function) @ hot.js?hash=d34b0a8…:1
hot.js:121 Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined.(…)
```

**Comment by sammkj** [comment]
I'm pretty sure I have seen this error before. Can you check and see if you're exporting the component? For example:

```
import React from 'react';

const componentA = () => (<div>Component A</div>);

export default componentA;   //----- sometimes I forgot to export, this was one of the cause
```

**Comment by rickmed** [comment]
@sammkj I am. Like this thought, export const compA = ...

**Comment by sammkj** [comment]
So you imported the component like this?

```
import { compA } from '/path/to/compA';
```

If this is not the cause of the problem, I guess I cannot help you further. Sorry! :(

**Comment by rickmed** [comment]
yeap: const { compA } = require('./compA.jsx') @sammkj

**Comment by gadicc** [comment]
These are usually always caused by the specified (or omitted) babel plugins/presets, *and their order*.  You might just have to play around.  I did once also see on the forums a case where an import on the preceding line was to an invalid file, and it seemed to cause this error somehow.  In any case, with Meteor 1.3.3 beta.0 out with official .babelrc support, I'll unfortunately no longer be supporting these kinds of issues.  But if you don't manage to resolve this, as an existing user, I'll take a look once meteor-hmr is working with 1.3.3.

**Comment by rickmed** [comment]
@gadicc I tried this setup 
```
{
  "presets": [
    "meteor",
    "es2015",
    "react"
  ],
  "plugins": [....]
}
```
and didn't work. Also, tried with meteor 1.3.3.beta.0 and it has issues with babel also (understandably since it is only the first beta release).
Would be great if meteor-hmr could make sure it works with <meteor 1.3.3 until it is officially release...

**Comment by gadicc** [comment]
Can you try removing `es2015`?  It conflicts on certain features with `meteor`, which already provides all the es2015 features.

**Comment by rickmed** [comment]
@gadicc so weird, now the same exact setup it works (with es2015), when running meteor it sometimes enables the ecmascript-hot package and sometimes not. EDIT: after it worked, I've tried running meteor a few times and ecmascript-hot is not loaded

**Comment by clayne11** [comment]
This is definitely a problem with `2.0.0-beta-3`. Using `beta-2` it works fine for me.

```
W20160526-16:50:08.615(-4)? (STDERR) /Users/clayne/workspace/r3ntr_meteor/.meteor/local/build/programs/server/packages/casalova_react-hot-loader-patch.js:25
W20160526-16:50:08.615(-4)? (STDERR) import 'react-hot-loader/patch';                                     // 1
W20160526-16:50:08.615(-4)? (STDERR) ^^^^^^
W20160526-16:50:08.617(-4)? (STDERR) SyntaxError: Unexpected reserved word
W20160526-16:50:08.617(-4)? (STDERR)     at /Users/clayne/workspace/r3ntr_meteor/.meteor/local/build/programs/server/boot.js:278:30
W20160526-16:50:08.617(-4)? (STDERR)     at Array.forEach (native)
W20160526-16:50:08.617(-4)? (STDERR)     at Function._.each._.forEach (/Users/clayne/.meteor/packages/meteor-tool/.1.3.2_4.lbyo5v++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/underscore/underscore.js:79:11)
W20160526-16:50:08.617(-4)? (STDERR)     at /Users/clayne/workspace/r3ntr_meteor/.meteor/local/build/programs/server/boot.js:133:5
```

**Comment by clayne11** [comment]
Now I'm not sure that `2.0.0-beta.3` is the culprit. I'm not sure what happened, but all of the sudden I'm completely unable to build my app and I keep getting this error. I have no idea what's wrong. It's very frustrating though and I've spent 1.5 days trying to fix this issue and I've made no progress on my work.

**Comment by tomitrescak** [comment]
This is probably Sihoť in the dark, but adding es2015 and meteor and react did the trick for me

**Comment by clayne11** [comment]
I've been using the exact same `.babelrc` for about 2 months and it all of the sudden stopped working. I now can't run any version of my project in any repo using and version of `gadicc:ecmascript-hot`.

**Comment by tomitrescak** [comment]
I had this:

```
{
  "presets": [
    "meteor
  ],
  "plugins": [ "react-hot-loader/babel" ]
}
```

Then I reset the project and changed to this:

```
{
  "presets": [
    "meteor",
    "es2015",
    "react"
  ],
  "plugins": [ "react-hot-loader/babel" ]
}
```

All started to be fine ...

**Comment by clayne11** [comment]
Looks like this might be the issue: https://github.com/meteor/babel-preset-meteor/issues/1

**Comment by gadicc** [comment]
Hey all, sorry, didn't realize this was affecting older versions.  My time until now has been spent updating meteor-hmr for Meteor 1.3.3, which I hoped would alleviate this issue; as it happens my new intimacy with Meteor 1.3.3, and your most recent comment (thanks!), might help solve this issue for older versions though (for 1.3.3 you could try running from `master`; it seems to work so far but I haven't had enough time testing it to feel comfortable making even a beta release).

What does `npm list babel-preset-meteor` show?  You want `v6.6.7` (or I guess `v6.6.8`), *not* a newer version, which you might have got from an `npm update`.  Unfortunately in `babel-compiler-babelrc` I was only checking for a minimum version, whereas I guess I should have pinned to an exact version.  The new version doesn't do the `import` transforms, as it relies on a [new, separate compiler](https://github.com/benjamn/reify) to do the transforms.  That's why `es2015` might help alleviate the issue, since it in this case, the conflict with meteor actually helps, but this isn't really how we want to solve this and I think it could cause bigger issues overall.

Another issue is that I wasn't invalidating the cache against changed versions of the meteor babel preset, which would further complicate this.  (Changing your .babelrc however is one way to invalidate the cache).

In short, if you have a different version, can you try:

1. `npm i --save-dev babel-preset-meteor@6.6.7`
1. `rm .meteor/local/plugin-cache/gadicc_ecmascript-hot/local/*` (in project root)

If that doesn't help you can try `rm -rf .meteor/local/{b*,p*}` too.

I'll try follow this issue over the weekend but I might not have a chance to sit down with it until Monday, sorry :(  But hopefully the above is the actual cause and fix for this.

**Comment by clayne11** [comment]
Pinning `babel-preset-meteor` to version `6.6.8` fixed the issue for me.

I swear to god this is one of the first things I tried though and then I proceeded to waste 2 entire work days trying to fix this issue. I guess I didn't do this but I could have sworn I did.

Anyways, that fixed the issue for me.

**Comment by gadicc** [comment]
Thanks for reporting back, happy this is no longer affecting you.  Unfortunately due to a small lack of foresight on my part, if you changed the babel-preset-meteor version without manually clearing the cache, you'd still end up using the cached version with the bad code.  I had a check in place for a minimum version, but didn't bank on on future versions breaking backwards compatibility... the meteor preset doesn't follow semver because meteor-babel always pins to an exact version of it.  So apologies for this.  Fortunately these kinds of things are no longer an issue for meteor-hmr moving forward, as we rely solely on the official babelrc support.

**Comment by gadicc** [comment]
Hey all, I think this is resolved, but anyone is welcome to re-open if they're still having issues not resolved by the above.

**Comment by semeano** [comment]
Hi! I'm new to Meteor and I'm getting the following error when running the app:

```
I20160715-12:39:27.637(1)? [gadicc:hot] modules-runtime - current: v0.6.5, supported v0.6.3; This might be ok but look out for newer versions and issues on meteor-hmr GitHub.
W20160715-12:39:28.822(1)? (STDERR)           
W20160715-12:39:28.824(1)? (STDERR) /Users/semeano/Development/mvp/app/.meteor/local/build/programs/server/app/app.js:15
W20160715-12:39:28.825(1)? (STDERR) import React from 'react';                                                    
W20160715-12:39:28.825(1)? (STDERR) ^^^^^^    
W20160715-12:39:28.830(1)? (STDERR) SyntaxError: Unexpected reserved word
W20160715-12:39:28.830(1)? (STDERR)     at /Users/semeano/Development/mvp/app/.meteor/local/build/programs/server/boot.js:292:30
W20160715-12:39:28.831(1)? (STDERR)     at Array.forEach (native)
W20160715-12:39:28.831(1)? (STDERR)     at Function._.each._.forEach (/Users/semeano/.meteor/packages/meteor-tool/.1.3.4_4.rwnnka++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/underscore/underscore.js:79:11)
W20160715-12:39:28.831(1)? (STDERR)     at /Users/semeano/Development/mvp/app/.meteor/local/build/programs/server/boot.js:133:5
=> Exited with code: 8
```

Could it be because I'm using version 0.6.5 of modules-runtime?

Thanks.

**Comment by gadicc** [comment]
See if `npm i --save-dev babel-preset-meteor` fixes the problem (to update your version of this package to the latest).

You can also check if this works without `meteor-hmr`, because it affected a few releases of Meteor too.  (I guess just check if you're using the latest Meteor).

**Comment by semeano** [comment]
I have Meteor v1.3.4.4. I also tried running `rm -rf .meteor/local/{b*,p*}` to no avail...
